### ISSUE TITLE
The Problem of Genetic Algorithm for Predicting Neural Networks

### ISSUE THREAD

**Issue by Spaereaude** [Spaereaude/Genetic-Algorithm-Problems#1]
During the process of running genetic algorithm, the results are all NaN, but there are no illegal values in the source file, and the function graph of loss value and iteration number is also normal. Why is this?

```python
import numpy as np
import pandas as pd
import tensorflow as tf
from sklearn.model_selection import train_test_split
from geneticalgorithm import geneticalgorithm as ga
from tensorflow.keras import layers
 
data = pd.read_excel('D:/test.xlsx', header=None)
 
X = data.iloc[:, :-1]  # input parameters
y = data.iloc[:, -1]  # output parameters
 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
 
# Define constraint rules (set the definition field, you can ignore this code, just know that there are discrete regions and continuous definition fields in the input parameters)
def check_constraints(params):
    # Constraints on columns 1 to 9 (0 index) (discrete domain)
    discrete_bounds = [
        [0, 0.67, 1, 1.2, 999],
        [0, 1],
        [0, 0.67, 1, 1.2, 999],
        [0, 1],
        [500, 1000, 2000, 3000, 7000],
        [0, 13, 15, 20, 24, 28, 36],
        [0.67, 1.2, 2],
        [1.5, 2, 3],
        [1000, 2000, 7000, 12000]
    ]
    for i in range(9):
        if params[i] not in discrete_bounds[i]:
            return False
 
     # Constraints on columns 10 to 40 (0-index) (continuous definition domain)
     if np.any(params[9:33] < 0) or np.any(params[9:33] > 100):
         return False

    return True
 
# Define fitness function
def fitness_function(params):
    if not check_constraints(params):
        return 1e9  # Penalty for violating constraints
 
    input_params = np.array(params).reshape(1, -1)
    output = model.predict(input_params).flatten()[0]
 
    if np.isnan(output):
        print("Warning: NaN encountered in fitness function")
        print("Input parameters:", params)
        return 1e9  # Penalty for NaN values
 
    if output <= 0 or output > 100:  # Output Range Constraint
        return 1e9  # Penalty for violating constraints
    else:
        return output
 
def on_generation_callback(ga_instance):
    generation = ga_instance.generations_completed
    best_solution = ga_instance.best_solution
    best_fitness = ga_instance.best_function
    print(f"Generation {generation}: Best solution = {best_solution}, Best fitness = {best_fitness}")
    
# Define genetic algorithm parameters
algorithm_parameters = {
    'max_num_iteration': 4000,
    'population_size': 50,
    'mutation_probability': 0.1,
    'elit_ratio': 0.01,
    'crossover_probability': 0.5,
    'parents_portion': 0.3,
    'crossover_type': 'uniform',
    'max_iteration_without_improv': None
}
 
# Establishing a neural network model
model = tf.keras.Sequential()
ANN_number = 50
model.add(tf.keras.layers.Dense(ANN_number, input_shape=[X.shape[1],], activation='relu'))
model.add(tf.keras.layers.Dense(ANN_number, activation='relu'))
model.add(tf.keras.layers.Dense(ANN_number, activation='relu'))
model.add(layers.Dropout(0.2))
model.add(tf.keras.layers.Dense(1))
# model.summary()    # View basic information of the model
model.compile(optimizer='adam', loss='mse')    # Optimize neural network model settings
history = model.fit(X_train, y_train, epochs=100, verbose=0)
 
# Define variable boundaries for genetic algorithms
varbound = np.vstack((np.array([-np.inf]*40), np.array([np.inf]*40))).T
 
model_genetic = ga(function=fitness_function, dimension=40,
                    variable_type='real', variable_boundaries=varbound,
                    algorithm_parameters=algorithm_parameters)
 
for generation in range(algorithm_parameters['max_num_iteration']):
    model_genetic.run()
    best_solution = model_genetic.best_variable
    best_fitness = model_genetic.best_function
    print(f"Generation {generation+1}: Best solution = {best_solution}, Best fitness = {best_fitness}")
```
The following are the results of the operation:
![微信图片_20230530121738](https://github.com/Spaereaude/Genetic-Algorithm-Problems/assets/63953011/5db90923-8d56-4350-ab35-f5873d31a3a4)
